### DISCUSSION TITLE
Site API / API key

### DISCUSSION THREAD

**Discussion by evanlegamer** [plausible/analytics#1460] [Self-Hosted Support] [answered]
### Past Issues Searched

- [X] I have searched open and closed issues to make sure that the bug has not yet been reported

### Issue is a Bug Report

- [X] This is a bug report and not a feature request, nor asking for self-hosted support

### Describe the bug

I created my API keys in the Plausible dashboard. I'm using the latest version 1.4.0.

After created my keys, I can access to Stats API but not Site API

I have
```json
{
    "error": "Invalid API key. Please make sure you're using a valid API key with access to the resource you've requested."
}
```

### Expected behavior

Access to the Site API

### Screenshots

<img width="422" alt="Capture d’écran 2021-11-09 165555" src="https://user-images.githubusercontent.com/18480367/140959399-72f0c3ff-ea0d-445e-b5c9-3608dafe1e3b.png">

<img width="1248" alt="Capture d’écran 2021-11-09 165648" src="https://user-images.githubusercontent.com/18480367/140959416-cff43484-0e65-4e72-b23d-f5f5bdfff920.png">

### Environment

```markdown
- OS: My pc is on windows, server is on Linux Ubuntu
- Browser: Chrome
- Browser Version: 95.0.4638.69 (Official Build) (64-bit)
```

**Comment by ukutaht** [comment, marked as answer]
Only API keys for the Stats API can be created through the UI. We currently provision keys for the Sites API manually.

It is possible to insert an API key directly in to the database. It needs to have scope `sites:provision:*`

**Reply by evanlegamer** [reply]
Okay, i'm really interested to do that, how can i login to the database and add the key ?

**Reply by ukutaht** [reply]
Details on how to connect to the database: https://github.com/plausible/analytics/discussions/621#discussioncomment-293169

You can then look at the `api_keys` table and figure out how to insert the key. I cannot provide the full insert statement at the moment.

**Reply by evanlegamer** [reply]
Ok i succeed to do it

I connected to the database and then

```
UPDATE api_keys SET scopes = '{sites:provision:*}' WHERE name = 'Development';
```

**Reply by ukutaht** [reply]
Sweet! Enjoy Plausible

**Reply by 1kevinson** [reply]
> Ok i succeed to do it
> 
> I connected to the database and then
> 
> ```
> UPDATE api_keys SET scopes = '{sites:provision:*}' WHERE name = 'Development';
> ```

I did the same but got the same error

Edit : finally it worked 👨🏽‍💻😁

**Reply by sptv2reg** [reply]
**Edit: Found the answer in the documentation (Please [contact us](https://plausible.io/contact) to get an API key with permissions for the endpoints listed in this document)**

Is the sites API available on hosted?

**Reply by 1kevinson** [reply]
> Is the sites API available on hosted?
Yes it's available 
Just use your custom url instead of plausible.io when you want to perform requests

**Reply by shashwatadi** [reply]
Hi, can you share the steps to add API key to the db? Thanks

**Comment by mattvb91** [comment]
@evanlegamer do you have any references for how you generated the key? does it have to be generated with the SECRET_KEY_BASE that is set through the env? Also not sure of the key_prefix if that needs to be set

**Reply by rdvo** [reply]
yeah +1 here, If I want to utilize the API on my own self hosted instance, what do I need to do? Can someone help. thanks

**Reply by mattvb91** [reply]
I ended up writing a connector to the database directly and perform manipulations I need directly. Pitty we wont get good API access on the open source version but it goes against their business model.

**Comment by Donnyvdm** [comment]
Okay, I had this problem too and managed to solve it using the pointers in this issue, but writing the correct query took a little time, so hopefully I can help someone else to make it quicker.

Assuming you want to add the `sites:provision:*`-scope to an existing api key, make sure you know the key name and use it in the query below:
```{SQL}
UPDATE api_keys SET scopes = array_append(scopes, 'sites:provision:*') WHERE name='my-key';
```
Alternatively, if you want to use an api key only for site provisioning (and remove the default access to the stats API), you can use the query by evanlegamer https://github.com/plausible/analytics/discussions/1460#discussioncomment-1618889 :
```{SQL}
UPDATE api_keys SET scopes = '{sites:provision:*}' WHERE name='my-key';
```

If you're unsure how/where to run this query, here is a oneliner for the default self-hosted setup. For this to work, you need to find the container ID or name of the docker container that is running Plausible's postgres database. In my setup it's called `plausible_db`. Then run the query as follows:

```
docker exec plausible_db psql -U postgres -d plausible_db -c "UPDATE api_keys SET scopes = array_append(scopes, 'sites:provision:*') WHERE name='my-key';"
```

**Comment by johannbuscail** [comment]
It's been nearly 3 years and the api access is still terrible. Why ?

**Comment by chrisdeeming** [comment]
Yeah it's really confusing to me why we're having to modify the database directly in order to create an API key for provisioning. Even worse, it's really surely just a case of having a few checkboxes in order to specify the scopes for each key? I'm enjoying Plausible so far, but this was disappointing and, honestly, far too lacklustre compared to everything else you've built here.

**Comment by ahsanbod** [comment]
/api/v1/sites takes me to 404. 
Is it available for self hosted version?

**Comment by AquaMCU** [comment]
I have the same issue: /api/v2/query is giving me a 404 as well

I tested in 2.x as well as in 3.0.1 self hosted docker version
